subsys/shell/shell.c

```diff
@@ -1321,6 +1321,7 @@ void shell_fprintf(const struct shell *shell, enum shell_vt100_color color,
 	va_list args = { 0 };
 
 	if (k_current_get() != shell->ctx->tid) {
+printk("shell: hehe, I'm deadlocking!\n");
```

Comment:
Yes this is problem.
Current mutex implementation in shell assumes that all the access to transport medium is synchronized by shell.

Reply:
> Current mutex implementation in shell assumes that all the access to transport medium is synchronized by shell.

If only that! It assumes that if shell does something from thread no.1, that request to print something from thread no.2 can be blocked forever. Which doesn't make sense, e.g. if thread no.1 itself blocked on thread no.2 completion.